### PR TITLE
173 - Use of references inside codeList

### PR DESCRIPTION
**Pull request by mauroz77**
- Add support for references in codeList.
- Add support to references within references.

### CODE REVIEW

File: test/functional/fixtures/references/references_within_references/input.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+					"restrictions": {}
+				}
+			]
+		}
+	],
+	"references": {
+		"ID_REG_EXP": "^[\\w\\s\\W]{5,}$",
+		"MALE": "Male",
+		"FEMALE": "Female",
+		"SEX": ["#/MALE", "#/FEMALE"]
```
Comment (joneubank) [inline test/functional/fixtures/references/references_within_references/input.json:45]: Did we cover the case where the nested reference is also an array?

```json
"LETTERS": ["A","B"],
"NESTED_REFERENCE": ["#/LETTERS", "C"]
```

File: test/functional/fixtures/references/cyclic_references/input.json [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+				}
+			]
+		}
+	],
+	"references": {
+		"ID_REG_EXP": "^[\\w\\s\\W]{5,}$",
+		"MALE": "Male",
+		"FEMALE": "Female",
+		"OTHER": "#/UNKNOWN",
+		"UNKNOWN": "#/OTHER",
```
Comment (joneubank) [inline test/functional/fixtures/references/cyclic_references/input.json:46]: 👍 

Should we also test for a reference that references itself?

```json
"SELF_REFERENCE": "#/SELF_REFERENCE"
```

**Review by joneubank — CHANGES_REQUESTED** [review]
Looks great!

Reading through I recognized two more edge cases to test. Can we add those?

File: test/functional/referencesTest.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+const simpleReferencesOutput = require('./fixtures/references/simple_references/output.json') as DictionaryDocument;
+const codeListReferencesInput = require('./fixtures/references/codeList_references/input.json') as DictionaryDocument;
+const codeListReferencesOutput = require('./fixtures/references/codeList_references/output.json') as DictionaryDocument;
+const referencesWithinReferencesInput = require('./fixtures/references/references_within_references/input.json') as DictionaryDocument;
+const referencesWithinReferencesOutput = require('./fixtures/references/references_within_references/output.json') as DictionaryDocument;
+const nonExistingReferencesInput = require('./fixtures/references/non_existing_references/input.json') as DictionaryDocument;
+const cyclicReferencesInput = require('./fixtures/references/cyclic_references/input.json') as DictionaryDocument;
+
+describe('Replace references in the schemas of a dictionary', () => {
+    it('Should return the same original schema if dictionaty does not contain a references section', () => {
```
Comment (joneubank) [inline test/functional/referencesTest.ts:19]: `doctionaty` typo appears a couple times ;)

File: test/functional/fixtures/references/references_within_references/input.json (hunk 1 above)
Comment (mauroz77) [reply test/functional/fixtures/references/references_within_references/input.json]: Changed input.json to reflect this scenario.
```
"UNKNOWN": ["Unknown", "No answer"],
"SEX": ["#/MALE", "#/FEMALE", "#/UNKNOWN"]
```

File: test/functional/fixtures/references/cyclic_references/input.json (hunk 2 above)
Comment (mauroz77) [reply test/functional/fixtures/references/cyclic_references/input.json:46]: Added fixtures/references/self_references/input.json and tes case to handle this scenario:

```
"restrictions": {
    "codeList": ["#/SEX", "Other", "#/SELF_REFERENCE"]
}
```

```
"references": {
  "ID_REG_EXP": "^[\\w\\s\\W]{5,}$",
  "MALE": "Male",
  "FEMALE": "Female",
  "SELF_REFERENCE": "#/SELF_REFERENCE",
  "SEX": ["#/MALE", "#/FEMALE"]
}
```

File: test/functional/referencesTest.ts (hunk 3 above)
Comment (mauroz77) [reply test/functional/referencesTest.ts]: Done.

**Review by joneubank — APPROVED** [review]
Thanks!
